lucene/core/src/java/org/apache/lucene/search/LRUQueryCache.java

```diff
@@ -807,14 +896,35 @@ public BulkScorer bulkScorer(LeafReaderContext context) throws IOException {
       DocIdSet docIdSet;
       try {
         docIdSet = get(in.getQuery(), cacheHelper);
+      } catch (AsyncQueryLoadInProgressException e) {
+        // Query is being loaded asynchronously. Use uncached version but
+        // do not do a new load of the same query into the cache
+        return in.bulkScorer(context);
       } finally {
         lock.unlock();
       }
 
       if (docIdSet == null) {
         if (policy.shouldCache(in.getQuery())) {
-          docIdSet = cache(context);
-          putIfAbsent(in.getQuery(), docIdSet, cacheHelper);
+          // If asynchronous caching is requested, perform the same and return
+          // the uncached iterator
+          if (executor != null) {
+            FutureTask<Void> task = new FutureTask<>(() -> {
+              DocIdSet localDocIdSet = cache(context);
+              putIfAbsent(in.getQuery(), localDocIdSet, cacheHelper);
+
+              //remove the key from inflight -- the key is loaded now
+              inFlightAsyncLoadQueries.remove(in.getQuery());
```

Comment:
Disappointing that all this code is duplicated between `scorerSupplier` and `bulkScorer` methods ... can we somehow factor to share?

Reply:
Good point -- refactored to a common method.